lib/IRGen/IRGenSIL.cpp

```diff
@@ -2059,33 +2102,126 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+          // TODO: confirm unit test coverage for float-typed attr.
+          setAttrFn = IGM.getTFE_OpSetAttrFloatFn();
+          auto value = attr.value.getFloatValue();
+          LLVM_DEBUG(llvm::dbgs() << " Setting float-typed attr " << attrName
+                                  << " with value " << attr.value << ".\n");
+          // TensorFlow only supports 32-bit float attributes.  If we got a 16
+          // or 64 bit one, convert it to float.
+          bool losesInfo = false;
+          value.convert(APFloat::IEEEsingle(), APFloat::rmNearestTiesToEven,
+                        &losesInfo);
+          attrVal = llvm::ConstantFP::get(IGM.FloatTy, value.convertToFloat());
+        } else {
+          auto value = attr.value.getIntegerValue();
+          LLVM_DEBUG(llvm::dbgs() << " Setting int-typed attr " << attrName
+                                  << " with value " << attr.value << ".\n");
+          if (value.getBitWidth() == 1) {
+            setAttrFn = IGM.getTFE_OpSetAttrBoolFn();
+            attrVal =
+                llvm::ConstantInt::get(IGM.Int1Ty, value.isAllOnesValue());
+          } else {
+            setAttrFn = IGM.getTFE_OpSetAttrIntFn();
+            attrVal =
+                llvm::ConstantInt::get(IGM.Int64Ty, value.getLimitedValue());
+          }
+        }
+        Builder.CreateCall(setAttrFn, {op, attrNameAddr, attrVal});
+        break;
+      }
+      case SymbolicValue::String: {
+        auto attrValue = attr.value.getStringValue();
+        if (attrName != "__device") {
+          auto attrNameAddr = createStringValAddr(IGM, attrName.c_str());
+          auto attrValAddr = createStringValAddr(IGM, attrValue);
+          auto attrLength =
+              llvm::ConstantInt::get(IGM.SizeTy, attrValue.size());
+          auto *setAttrFn = IGM.getTFE_OpSetAttrStringFn();
+          Builder.CreateCall(setAttrFn,
+                             {op, attrNameAddr, attrValAddr, attrLength});
+        } else {
           // In this case, just let eager pick a default device.
-          if (deviceStr == "ALL_DEVICES")
+          if (attrValue == ALL_DEVICES)
             break;
           auto *setDeviceFn = IGM.getTFE_OpSetDeviceFn();
-          auto device = createStringValAddr(IGM, deviceStr);
+          auto device = createStringValAddr(IGM, attrValue);
           Builder.CreateCall(setDeviceFn, {op, device, status});
           checkOk(status);
-        } else {
-          llvm_unreachable("TODO: implement this");
         }
         break;
+      }
       case SymbolicValue::Metatype:
         // Set up dtype attr as in:
         //   TFE_OpSetAttrType(op, "dtype", TF_FLOAT);
         auto *setAttrTypeFn = IGM.getTFE_OpSetAttrTypeFn();
         auto dtypeValAddr = createStringValAddr(IGM, attrName.c_str());
         dtypeAttr = convertSwiftTypeToTF(attr.value.getMetatypeValue());
+        LLVM_DEBUG(llvm::dbgs() << " Setting type-typed attr " << attrName
```

Comment:
Remove the leading space in the message?

Reply:
I intentionally added such indentation, to make the debugging msg a bit easier to read.